### PR TITLE
Fix Application Cert Expiry Validation

### PR DESCRIPTION
**Pull request by Shaaali**
### Proposed changes in this pull request

$subject in order to resolve the issue : https://github.com/wso2/product-is/issues/14241

### When should this PR be merged

Please do not merge the PR until an approval comment is posted in the thread. 

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java [hunk 1]
```diff
@@ -1924,6 +1930,19 @@ private void doPreAddApplicationChecks(ServiceProvider serviceProvider, String t
...
+                if (isCertificateExpired(cert)) {
+                    String msg = "Application with name: '" + appName + "' has an expired application certificate.";
+                    throw buildClientException(EXPIRED_CERTIFICATE, msg);
+                }
+            } catch (CertificateException e) {
+                String msg = "Application with name: '" + appName + "' has malformed application certificate.";
+                throw buildClientException(INVALID_REQUEST, msg);
+            }
+        }
+
```
Comment (senthalan) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:1945]: can't we call `validateApplicationCertificate` other than duplicating the code

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java [hunk 2]
```diff
@@ -2467,17 +2475,53 @@ private void updateApplicationPermissions(ServiceProvider updatedApp, String upd
...
                     "is malformed.";
             throw buildClientException(INVALID_REQUEST,
-                    String.format(error, updatedApp.getApplicationName(), tenantDomain));
+                    String.format(error, serviceProvider.getApplicationName(), tenantDomain));
+        }
+        if (StringUtils.isNotEmpty(serviceProvider.getCertificateContent())) {
+            try {
+                X509Certificate cert = extractCertificate(serviceProvider.getCertificateContent());
+                if (isCertificateExpired(cert)) {
+                    if (isCertificateExpired(cert)) {
```
Comment (senthalan) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:2491]: why we need a double check ?

**Review by senthalan — APPROVED** [review]

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java (hunk 2 above)
Comment (Shaaali) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java]: Removed the double check.

**Comment by senthalan** [comment]
Closing as the PR is in the open state for a while. Please feel free to reopen and merge if needed.

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java [hunk 3]
```diff
@@ -2467,15 +2475,49 @@ private void updateApplicationPermissions(ServiceProvider updatedApp, String upd
...
+                        "' has a malformed application certificate.";
+                throw buildClientException(INVALID_REQUEST, msg);
+            }
+        }
+    }
+
+    private X509Certificate extractCertificate(String certData) throws CertificateException {
+
+        byte[] bytes = Base64.decode(certData);
+        X509Certificate cert;
```
Comment (Shaaali) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:2506]: cert variable is redundant

**Pull request closed without merging**